flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/TemporalJoinITCase.scala

```diff
@@ -260,6 +393,214 @@ class TemporalJoinITCase(state: StateBackendMode)
+    val expected = List(
+      "+I(1,Euro,12,2020-08-15T00:01,114,2020-08-15T00:00:01)",
+      "+I(2,US Dollar,1,2020-08-15T00:02,102,2020-08-15T00:00:02)",
+      "-U(2,US Dollar,1,2020-08-16T00:03,106,2020-08-16T00:02)",
+      "+U(2,US Dollar,18,2020-08-16T00:03,106,2020-08-16T00:02)")
+    assertEquals(expected.sorted, rawResult.sorted)
+  }
+
+  @Test
+  def testEventTimeLeftTemporalJoin(): Unit = {
+    val sql = "INSERT INTO rowtime_default_sink " +
+      " SELECT o.order_id, o.currency, o.amount, o.order_time, r.rate, r.currency_time " +
+      " FROM orders_rowtime AS o LEFT JOIN versioned_currency_with_single_key " +
+      " FOR SYSTEM_TIME AS OF o.order_time as r " +
+      " ON o.currency = r.currency"
+    tEnv.executeSql(sql).await()
+
+    val rawResult = getRawResults("rowtime_default_sink")
+    val expected = List(
+      "+I(1,Euro,12,2020-08-15T00:01,114,2020-08-15T00:00:01)",
+      "+I(2,US Dollar,1,2020-08-15T00:02,102,2020-08-15T00:00:02)",
+      "+I(3,RMB,40,2020-08-15T00:03,702,2020-08-15T00:00:04)",
+      "+I(4,Euro,14,2020-08-16T00:04,118,2020-08-16T00:01)",
+      "-U(2,US Dollar,1,2020-08-16T00:03,106,2020-08-16T00:02)",
+      "+U(2,US Dollar,18,2020-08-16T00:03,106,2020-08-16T00:02)",
+      "+I(5,RMB,40,2020-08-16T00:03,null,null)",
+      "+I(6,RMB,40,2020-08-16T00:04,null,null)",
+      "-D(6,RMB,40,2020-08-16T00:04,null,null)")
+    assertEquals(expected.sorted, rawResult.sorted)
+  }
+
+  @Test
+  def testEventTimeTemporalJoinChangelogUsingBeforeTime(): Unit = {
+    val sql = "INSERT INTO rowtime_default_sink " +
+      " SELECT o.order_id, o.currency, o.amount, o.order_time, r.rate, r.currency_time " +
+      " FROM orders_rowtime AS o LEFT JOIN currency_using_update_before_time " +
+      " FOR SYSTEM_TIME AS OF o.order_time as r " +
+      " ON o.currency = r.currency"
+    tEnv.executeSql(sql).await()
+
+    val rawResult = getRawResults("rowtime_default_sink")
+    val expected = List(
+      "+I(1,Euro,12,2020-08-15T00:01,null,null)",
+      "+I(2,US Dollar,1,2020-08-15T00:02,102,2020-08-15T00:00:02)",
+      "+I(3,RMB,40,2020-08-15T00:03,702,2020-08-15T00:00:04)",
+      "+I(4,Euro,14,2020-08-16T00:04,118,2020-08-16T00:01)",
+      "-U(2,US Dollar,1,2020-08-16T00:03,106,2020-08-16T00:02)",
+      "+U(2,US Dollar,18,2020-08-16T00:03,106,2020-08-16T00:02)",
+      "+I(5,RMB,40,2020-08-16T00:03,null,null)",
+      "+I(6,RMB,40,2020-08-16T00:04,null,null)",
+      "-D(6,RMB,40,2020-08-16T00:04,null,null)")
+    assertEquals(expected.sorted, rawResult.sorted)
+  }
+
+  @Test
+  def testEventTimeLeftTemporalJoinUpsertSource(): Unit = {
+    // Note: The WatermarkAssigner of upsertSource is followed after ChangelogNormalize,
+    // when the parallelism > 1 and test data doesn't cover all parallelisms, it returns
+    // Long.MaxValue as final watermark until all parallelism finished.
+    // This may leads the test failed because the test data doesn't cover every parallelism.
```

Comment:
Add a comment `// TODO: remove the single parallelism once FLINK-19878 is fixed.`